spring-integration-event/src/main/java/org/springframework/integration/event/inbound/ApplicationEventListeningMessageProducer.java

```diff
@@ -116,20 +129,45 @@ public void onApplicationEvent(ApplicationEvent event) {
 		}
 	}
 
+	@Override
+	protected Object evaluatePayloadExpression(Object rootObject) {
+		if (!this.payloadExpressionSet && rootObject instanceof PayloadApplicationEvent) {
+			return ((PayloadApplicationEvent) rootObject).getPayload();
```

Comment:
> I don't think we expect users to publish arbitrary object as events.

The `eventPublisher` allows that directly (`ApplicationEventPublisherAware`) and our `ApplicationEventPublishingMessageHandler` has that ability already with the new `publishPayload` option.

My position for this "unwrapping" to be symmetrical with the POJO's `@EventListener`.

From other side I see your point and we really can just rely on the `payloadExpression` as `payload` from the end-user.
And emit as is `PayloadApplicationEvent` otherwise. Although with my change it would be a bit faster, because we don't need to evaluate expression for each payload event :smile: .In this case we can reach the whole `event` object with the expression like the `#root`, BTW...